File: app/src/test/java/li/klass/fhem/domain/WeatherDeviceTest.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.domain;

import org.junit.Test;

import java.util.List;

import li.klass.fhem.domain.core.DeviceXMLParsingBase;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.notNullValue;
import static org.hamcrest.core.Is.is;

public class WeatherDeviceTest extends DeviceXMLParsingBase {
    @Test
    public void testForCorrectlySetAttributes() {
        WeatherDevice device = getDefaultDevice();

        assertThat(device.getName(), is(DEFAULT_TEST_DEVICE_NAME));
        assertThat(device.getRoomConcatenated(), is(DEFAULT_TEST_ROOM_NAME));

        assertThat(device.getIcon(), is("cloudy.png"));
        assertThat(device.getCondition(), is("Bewölkt"));
        assertThat(device.getHumidity(), is("43 (%)"));
        assertThat(device.getTemperature(), is("19 (°C)"));
        assertThat(device.getWind(), is("NW mit 11 km/h"));
        assertThat(device.getState(), is("T: 19  H: 43  W: 11"));

        List<WeatherDevice.WeatherDeviceForecast> forecasts = device.getForecasts();
        assertThat(forecasts.size(), is(3));

        WeatherDevice.WeatherDeviceForecast firstForecast = forecasts.get(0);
        assertThat(firstForecast.getIcon(), is("chance_of_rain.png"));
        assertThat(firstForecast.getDayOfWeek(), is("Mi."));
        assertThat(firstForecast.getCondition(), is("Vereinzelt Regen"));
        assertThat(firstForecast.getDate(), is("2012-05-02"));
        assertThat(firstForecast.getHighTemperature(), is("20 (°C)"));
        assertThat(firstForecast.getLowTemperature(), is("9 (°C)"));

        WeatherDevice.WeatherDeviceForecast secondForecast = forecasts.get(1);
        assertThat(secondForecast.getIcon(), is("mostly_sunny.png"));
        assertThat(secondForecast.getDayOfWeek(), is("Do."));
        assertThat(secondForecast.getCondition(), is("Meist sonnig"));
        assertThat(secondForecast.getDate(), is("2012-05-03"));
        assertThat(secondForecast.getHighTemperature(), is("19 (°C)"));
        assertThat(secondForecast.getLowTemperature(), is("8 (°C)"));

        WeatherDevice.WeatherDeviceForecast thirdForecast = forecasts.get(2);
        assertThat(thirdForecast.getIcon(), is("sunny.png"));
        assertThat(thirdForecast.getDayOfWeek(), is("Fr."));
        assertThat(thirdForecast.getCondition(), is("Klar"));
        assertThat(thirdForecast.getDate(), is("2012-05-04"));
        assertThat(thirdForecast.getHighTemperature(), is("19 (°C)"));
        assertThat(thirdForecast.getLowTemperature(), is("10 (°C)"));

        assertThat(device.getSetList().getEntries().size(), is(0));

        assertThat(device.getLogDevices(), is(notNullValue()));
        assertThat(device.getDeviceCharts().size(), is(1));

        device.readICON("");
    }

    @Override
    protected String getFileName() {
        return "weather.xml";
    }
}

File: app/src/test/java/li/klass/fhem/domain/WeatherDeviceTest.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.domain

import li.klass.fhem.devices.backend.WeatherService
import li.klass.fhem.devices.backend.WeatherService.WeatherForecastInformation
import li.klass.fhem.domain.core.DeviceXMLParsingBase
import org.assertj.core.api.Assertions.assertThat
import org.joda.time.LocalDate
import org.junit.Test

class WeatherDeviceTest : DeviceXMLParsingBase() {
    @Test
    fun testForCorrectlySetAttributes() {
        val device = getDefaultDevice(GenericDevice::class.java)

        assertThat(device.name).isEqualTo(DeviceXMLParsingBase.DEFAULT_TEST_DEVICE_NAME)
        assertThat(device.roomConcatenated).isEqualTo(DeviceXMLParsingBase.DEFAULT_TEST_ROOM_NAME)

        assertThat(stateValueFor(device, "icon")).isEqualTo("cloudy")
        assertThat(stateValueFor(device, "condition")).isEqualTo("Bewölkt")
        assertThat(stateValueFor(device, "humidity")).isEqualTo("43.0 (%)")
        assertThat(stateValueFor(device, "temperature")).isEqualTo("19.0 (°C)")
        assertThat(stateValueFor(device, "wind_condition")).isEqualTo("NW mit 11 km/h")
        assertThat(device.state).isEqualTo("T: 19  H: 43  W: 11")

        val forecasts = WeatherService().forecastsFor(device)
        assertThat(forecasts).containsExactly(
                WeatherForecastInformation(
                        date = LocalDate.parse("2012-05-01"),
                        weekday = "Di.",
                        condition = "Bewölkt",
                        temperature = "19.0 (°C)",
                        icon = String.format(WeatherService.urlTemplate, "cloudy"),
                        humidity = "43.0 (%)"
                ),
                WeatherForecastInformation(
                        date = LocalDate.parse("2012-05-02"),
                        weekday = "Mi.",
                        condition = "Vereinzelt Regen",
                        temperature = "9 (°C) - 20 (°C)",
                        icon = String.format(WeatherService.urlTemplate, "chance_of_rain")
                ),
                WeatherForecastInformation(
                        date = LocalDate.parse("2012-05-03"),
                        weekday = "Do.",
                        condition = "Meist sonnig",
                        temperature = "8 (°C) - 19 (°C)",
                        icon = String.format(WeatherService.urlTemplate, "mostly_sunny")
                ),
                WeatherForecastInformation(
                        date = LocalDate.parse("2012-05-04"),
                        weekday = "Fr.",
                        condition = "Klar",
                        temperature = "10 (°C) - 19 (°C)",
                        icon = String.format(WeatherService.urlTemplate, "sunny")
                )
        )

        assertThat(device.xmlListDevice.setList.entries).isEmpty()
    }

    override fun getFileName(): String = "weather.xml"
}
